scanpy/external/tl/_trimap.py

```diff
@@ -0,0 +1,148 @@
+     archivePrefix = "arXiv",
+     eprint = {1910.00204},
+     year = 2019,
+    }
+
+    Input
+    ------
+
+    n_dims: Number of dimensions of the embedding (default = 2)
+
+    n_inliers: Number of inlier points for triplet constraints (default = 10)
+
+    n_outliers: Number of outlier points for triplet constraints (default = 5)
+
+    n_random: Number of random triplet constraints per point (default = 5)
+
+    metric: Distance measure ('euclidean' (default), 'manhattan', 'angular',
+    'hamming')
+
+    lr: Learning rate (default = 1000.0)
+
+    n_iters: Number of iterations (default = 400)
+
+    knn_tuple: Use the pre-computed nearest-neighbors information in form of a
+    tuple (knn_nbrs, knn_distances) (default = None)
+
+    apply_pca: Apply PCA to reduce the dimensions to 100 if necessary before the
+    nearest-neighbor calculation (default = True)
+
+    opt_method: Optimization method ('sd': steepest descent,  'momentum': GD
+    with momentum, 'dbd': GD with momentum delta-bar-delta (default))
+
+    verbose: Print the progress report (default = True)
+
+    weight_adj: Adjusting the weights using a non-linear transformation
+    (default = 500.0)
+
+    return_seq: Return the sequence of maps recorded every 10 iterations
+    (default = False)
+
+    Example
+    -------
+    
+    >>> import scanpy as sc
+    >>> import scanpy.external as sce
+    >>> pbmc = sc.datasets.pbmc68k_reduced()
+    >>> pbmc = sce.tl.trimap(pbmc, copy=True)
+    >>> sce.pl.trimap(pbmc, color=['bulk_labels'], s=10)
+
+    """
+
+    try:
+        from trimap import TRIMAP
+    except ImportError:
+        raise ImportError(
+            '\nplease install trimap: \n\n\tsudo pip install trimap'
+        )
+    adata = adata.copy() if copy else adata
+    start = logg.info('computing TRIMAP')
+
+    if isinstance(init_pos, str) and init_pos in adata.obsm.keys():
+        init_coords = adata.obsm[init_pos]
+    elif isinstance(init_pos, str) and init_pos == 'paga':
+        init_coords = get_init_pos_from_paga(adata, random_state=random_state)
+    else:
+        init_coords = init_pos  # Let TriMap handle it
+    if hasattr(init_coords, "dtype"):
+        init_coords = check_array(
+            init_coords, dtype=np.float32, accept_sparse=False
+        )
+
+    if 'X_pca' in adata.obsm:
+        n_dim_pca = adata.obsm['X_pca'].shape[1]
+        X = adata.obsm['X_pca'][:, : min(n_dim_pca, 100)]
+    else:
+        X = adata.X
+        if scp.issparse(X):
+            raise ValueError(
+                'trimap currently does not support sparse matrices. please'
+                'use a dense matrix or apply pca first.'
+            )
+    if metric not in ['euclidean', 'manhattan', 'angular', 'hamming']:
+        raise ValueError(
+            'metric \'' + metric + '\' is currently not supported '
+            'by trimap. acceptable values are \'euclidean\', \'manhattan\', '
+            '\'angular\', and \'hamming\'.'
```

Comment:
please avoid unnecessary escaping and prefer f-strings to concatenating strings with `+`:

```py
f'Metric {metric!r} is currently not supported by trimap. '
'Acceptable values are {", ".join(map(repr, allowed_metrics))}'
```